torchao/dtypes/uintx/csr_layout.py

```diff
@@ -0,0 +1,266 @@
+TorchAO.
+
+Key pieces:
+* ``CSRLayout`` – a stateless ``Layout`` subclass signalling that a tensor is
+  stored in CSR form.
+* ``CSR_AQTTensorImpl`` – a ``TensorImpl`` that actually holds the compressed
+  data and quantization parameters (scale / zero-point).
+"""
+
+from __future__ import annotations
+
+from dataclasses import dataclass
+from typing import Optional
+
+import torch
+from torch.utils._python_dispatch import return_and_correct_aliasing
+
+from torchao.dtypes.affine_quantized_tensor import (
+    AffineQuantizedTensor,
+    register_layout,
+)
+from torchao.dtypes.uintx.plain_layout import (
+    PlainAQTTensorImpl,
+    _aqt_is_int8_reduced_range,
+)
+from torchao.dtypes.utils import Layout, PlainLayout
+
+aten = torch.ops.aten
+
+# -----------------------------------------------------------------------------
+# Op-level helpers – dispatched during aten.linear if both inputs satisfy the
+# predicate.  In the first version we delegate to ``torch.sparse.mm``; follow-up
+# patches may swap in vendor SpMM kernels (FBGEMM / oneDNN).
+# -----------------------------------------------------------------------------
+
+
+def _linear_int8_act_int8_weight_csr_sparse_check(
+    input_tensor: torch.Tensor,
+    weight_tensor: torch.Tensor,
+    bias: Optional[torch.Tensor],
+) -> bool:
+    """Check if we can execute csr-optimised path."""
+    return (
+        isinstance(input_tensor, AffineQuantizedTensor)
+        and _aqt_is_int8_reduced_range(input_tensor)
+        and isinstance(weight_tensor, AffineQuantizedTensor)
+        and input_tensor.dtype == weight_tensor.dtype
+        and isinstance(input_tensor._layout, PlainLayout)
+        and isinstance(weight_tensor._layout, CSRLayout)
+    )
+
+
+def _linear_int8_act_int8_weight_csr_sparse_impl(
+    input_tensor: AffineQuantizedTensor,
+    weight_tensor: AffineQuantizedTensor,
+    bias: Optional[torch.Tensor],
+):
+    """Reference implementation using ``torch.sparse.mm``.
+
+    * ``input_tensor`` is **plain** INT8 AQT with per-tensor scale.
+    * ``weight_tensor`` is **CSR** INT8 AQT – compressed row representation.
+    * Produces *fp32* output multiplied by both quant scales, then casts back
+      to original activation dtype.
+    """
+    x_vals_int8 = input_tensor.tensor_impl.int_data
+    x_scale = input_tensor.tensor_impl.scale  # shape: (1,)
+
+    # The weight CSR tensor is stored as ``int_data`` in CSR layout.
+    w_csr = weight_tensor.tensor_impl.int_data  # 2-D sparse CSR INT8
+    w_scale = weight_tensor.tensor_impl.scale  # shape: (1,) or (row,)
+
+    # Reshape activations to 2-D (batch*seqlen, in_features)
+    x2d = x_vals_int8.reshape(-1, x_vals_int8.shape[-1]).to(torch.int32)
+
+    # As for now we don't have any kernel for spmm form int8 computation for ARM so we are upscaling it to fp32 for computation
+    # Once we intigrate the kernel in torch then we can use that kernel instead of torch.mm()
+    y_int32 = torch.mm(
```

Comment:
nit: I think the var names here could be a bit clearer - why is this y_int32? torch.mm should output float32 right?